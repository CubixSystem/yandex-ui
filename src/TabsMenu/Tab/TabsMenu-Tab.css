.TabsMenu-Tab {
    cursor: pointer;
    user-select: none;
    white-space: nowrap;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
}

.TabsMenu-Tab_disabled {
    cursor: default;
}

.TabsMenu-Tab .Link {
    position: relative;

    display: inline-block;
}
