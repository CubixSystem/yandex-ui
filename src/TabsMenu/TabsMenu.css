.TabsMenu {
    margin: 0;
    padding: 0;

    font-family: var(--tabs-menu-font-family);
}
