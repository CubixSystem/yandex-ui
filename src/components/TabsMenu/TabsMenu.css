.TabsMenu {
    margin: 0;
    padding: 0;

    font-family: var(--control-font-family);
}
