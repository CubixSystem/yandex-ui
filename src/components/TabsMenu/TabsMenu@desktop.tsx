import '../../polyfills/pointerfocus';

export * from './TabsMenu';
