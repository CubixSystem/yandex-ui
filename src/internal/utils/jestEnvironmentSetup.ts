/**
 * Возвращает DOM-узел для монтирования компонента.
 */
export const getRoot = (): HTMLElement | null => {
    return document.getElementById('root');
};

/**
 * Хелпер, создающий корневой контейнер для монтирования,
 * необходим для того, чтобы избежать ошибок при монтировании в enzyme.
 */
export const jestEnvironmentSetup = (): void => {
    // eslint-disable-next-line no-undef
    beforeEach(() => {
        const node = document.createElement('div');
        node.setAttribute('id', 'root');
        document.body.appendChild(node);
    });

    // eslint-disable-next-line no-undef
    afterEach(() => {
        const node = getRoot();
        if (node !== null) {
            document.body.removeChild(node);
        }
    });
};
