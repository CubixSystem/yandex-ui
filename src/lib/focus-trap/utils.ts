import { getTabbables } from './tabbable';

const FOCUS_TRAP_GUARD_ATTR = 'focusTrapGuard';

interface ScrollableElement {
    element: HTMLElement;
    scrollTop: number;
    scrollLeft: number;
}

let supportsPreventScrollCached: boolean | null = null;
function supportsPreventScroll() {
    if (supportsPreventScrollCached === null) {
        supportsPreventScrollCached = false;

        try {
            const element = document.createElement('div');

            element.focus({
                get preventScroll() {
                    supportsPreventScrollCached = true;
                    return true;
                },
            });
        } catch (e) {
            // ignore
        }
    }

    return supportsPreventScrollCached;
}

function getScrollableElements(element: HTMLElement): ScrollableElement[] {
    const scrollableElements: ScrollableElement[] = [];
    const rootScrollingElement = document.scrollingElement || document.documentElement;

    let parent = element.parentNode;
    while (parent instanceof HTMLElement && parent !== rootScrollingElement) {
        if (parent.offsetHeight < parent.scrollHeight || parent.offsetWidth < parent.scrollWidth) {
            scrollableElements.push({
                element: parent,
                scrollTop: parent.scrollTop,
                scrollLeft: parent.scrollLeft,
            });
        }
        parent = parent.parentNode;
    }

    if (rootScrollingElement instanceof HTMLElement) {
        scrollableElements.push({
            element: rootScrollingElement,
            scrollTop: rootScrollingElement.scrollTop,
            scrollLeft: rootScrollingElement.scrollLeft,
        });
    }

    return scrollableElements;
}

function restoreScrollPosition(scrollableElements: ScrollableElement[]) {
    for (const { element, scrollTop, scrollLeft } of scrollableElements) {
        element.scrollTop = scrollTop;
        element.scrollLeft = scrollLeft;
    }
}

export function getActiveElement() {
    return document.activeElement as HTMLElement | null;
}

export function isElementInScope(scope: Element, element: Element) {
    return scope.contains(element);
}

export function isFocusFree() {
    return getActiveElement() === document.body;
}

export function createGuard(tabIndex: number) {
    const element = document.createElement('span');
    element.tabIndex = tabIndex;
    Object.assign(element.style, {
        width: '1px',
        height: 0,
        padding: 0,
        overflow: 'hidden',
        position: 'fixed',
        top: '1px',
        left: '1px',
    });
    element.setAttribute('aria-hidden', 'true');
    element.dataset[FOCUS_TRAP_GUARD_ATTR] = 'true';

    return element;
}

export function isGuard(node: HTMLElement) {
    return Boolean(node.dataset[FOCUS_TRAP_GUARD_ATTR]);
}

export function removeElementFromDocument(element: Element) {
    if (!element.parentNode) {
        return;
    }

    element.parentNode.removeChild(element);
}

function focusWithoutScrolling(element: HTMLElement) {
    if (supportsPreventScroll()) {
        element.focus({ preventScroll: true });
    } else {
        const scrollableElements = getScrollableElements(element);
        element.focus();
        restoreScrollPosition(scrollableElements);
    }
}

export function focusElement(element?: HTMLElement | null, preventScroll?: boolean) {
    if (!element) {
        return;
    }

    try {
        if (preventScroll) {
            focusWithoutScrolling(element);
        } else {
            element.focus();
        }
    } catch (e) {
        // ignore
    }
}

export function focusFirstIn(scope: HTMLElement) {
    const [first] = getTabbables(scope, (node) => !isGuard(node));

    focusElement(first);
}

export function getNextTabbableIndex(current: number, count: number, step: number) {
    if (current === -1) {
        return 0;
    }

    return count > 0 ? (count + current + step) % count : -1;
}
