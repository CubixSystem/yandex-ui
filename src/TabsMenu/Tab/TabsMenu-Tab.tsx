/* eslint-disable no-nested-ternary */
import React, { useCallback, RefObject, ReactNode, KeyboardEventHandler, FC, MouseEventHandler } from 'react';

import { cnTabsMenu } from '../TabsMenu';
import './TabsMenu-Tab.css';

export interface ITabsMenuTabProps {
    /**
     * Идентификатор пункта меню
     */
    id: string;

    /**
     * Активный пункт меню
     */
    active?: boolean;

    /**
     * Недоступный для выбора пункт меню
     */
    disabled?: boolean;

    /**
     * Первый пункт меню
     *
     * @internal
     */
    first?: boolean;

    /**
     * Содержимое пункта меню. Например: текст или ссылка
     */
    content?: ReactNode;

    /**
     * Обработчик, вызываемый при клике на пункт меню
     */
    onClick?: MouseEventHandler<HTMLLIElement>;

    /**
     * @internal
     */
    innerRef?: RefObject<HTMLLIElement>;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * HTML атрибут `role`
     */
    role?: string;

    /**
     * HTML атрибут `tabIndex`
     */
    tabIndex?: number;

    /**
     * Обработчик события `onKeyDown`
     */
    onKeyDown?: KeyboardEventHandler<HTMLElement>;
}

interface ITabsMenuTabInternalProps extends ITabsMenuTabProps {
    onInternalClick?: (id: string) => void;
}

const Tab: FC<ITabsMenuTabInternalProps> = (props) => {
    const {
        id,
        innerRef,
        active,
        className,
        content,
        disabled,
        first,
        onClick,
        onInternalClick,
        ...otherProps
    } = props;

    const handleClick = useCallback((e: React.MouseEvent<HTMLLIElement>) => {
        onClick && onClick(e);
        onInternalClick && onInternalClick(id);
        // TODO ISL-10952: разобраться почему не все необходимые переменные указаны в deps и убрать игнор
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [id, onClick]);

    return (
        <li
            {...otherProps}
            aria-selected={active}
            ref={innerRef}
            className={cnTabsMenu(
                'Tab',
                {
                    active,
                    disabled,
                    first,
                },
                [className],
            )}
            role="tab"
            // На активном пункте устанавливаем tabIndex 0, чтобы иметь возможность
            // сфокусироваться с клавиатуры именно в него, у остальных пунктов устанавливаем -1,
            // чтобы они не участвовали в навигации.
            tabIndex={disabled ? undefined : active ? 0 : -1}
            onClick={disabled ? undefined : handleClick}
        >
            {content}
        </li>
    );
};

export const TabsMenuTab = React.memo(Tab);
