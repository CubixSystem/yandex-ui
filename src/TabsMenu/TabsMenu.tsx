import React, { KeyboardEvent, RefObject, useCallback, FC, ReactNode, createRef, useMemo } from 'react';
import { cn } from '@bem-react/classname';

import { isKeyCode, Keys } from '../lib/keyboard';
import { ITabsMenuTabProps, TabsMenuTab } from './Tab/TabsMenu-Tab';
import './TabsMenu.css';

export interface ITabsMenuProps {
    /**
     * Идентификатор активного пункта меню.
     */
    activeTab?: string;

    /**
     * Функция, меняющая активный пункт меню
     */
    onChange?: (tabId: string) => void;

    /**
     * Массив пунктов меню.
     */
    tabs: ITabsMenuTabProps[];

    /**
     * Массив с Ref пунктов меню
     *
     * @internal
     */
    tabsRefs?: RefObject<HTMLLIElement>[];

    /**
     * Дополнительные табы (используется в модификаторе withAdaptive)
     */
    addonAfter?: ReactNode;

    /**
     * Ссылка на корневой DOM элемент компонента.
     */
    innerRef?: RefObject<HTMLUListElement>;

    /**
     * Расположение пунктов меню (используется для доступности).
     *
     * @internal
     * @default 'vertical'
     */
    orientation?: 'horizontal' | 'vertical';

    /**
     * Дополнительный класс.
     */
    className?: string;
}

export const cnTabsMenu = cn('TabsMenu');

/**
 * Компонент для создания горизонтального меню.
 * @param {ITabsMenuProps} props
 */
export const TabsMenu: FC<ITabsMenuProps> = ({
    activeTab,
    onChange,
    className,
    innerRef,
    tabs,
    addonAfter,
    orientation = 'vertical',
    tabsRefs: externalTabsRefs,
    // @ts-ignore
    theme: _theme,
    // @ts-ignore
    layout: _layout,
    ...props
}) => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const tabsRefs = externalTabsRefs || useMemo(() => tabs.map(() => createRef<HTMLLIElement>()), [tabs]);

    const onKeyDown = useCallback(
        (event: KeyboardEvent) => {
            if (isKeyCode(event.keyCode, [Keys.LEFT, Keys.DOWN, Keys.RIGHT, Keys.UP])) {
                event.preventDefault();

                let nextTabMenuIndex = 0;
                const direction = isKeyCode(event.keyCode, [Keys.LEFT, Keys.UP]) ? -1 : 1;

                for (let i = 0; i < tabs.length; i++) {
                    if (tabs[i].id === activeTab) {
                        nextTabMenuIndex = i;
                        break;
                    }
                }

                for (let i = nextTabMenuIndex; i < tabs.length; i += direction) {
                    nextTabMenuIndex += direction;
                    // При необходимости тут можно реализовать цикличный выбор табов,
                    // при попадании индекса за границу списка,
                    // выставлять nextTabMenuIndex в противоположное значение.
                    const isEdge = nextTabMenuIndex >= tabs.length || nextTabMenuIndex < 0;
                    const isTabNonDisabled = tabs[nextTabMenuIndex] && !tabs[nextTabMenuIndex].disabled;
                    if (isEdge || isTabNonDisabled) {
                        break;
                    }
                }

                const nextTabMenu = tabs[nextTabMenuIndex];
                const nextTabMenuRef = tabsRefs[nextTabMenuIndex];

                if (
                    nextTabMenu !== undefined &&
                    nextTabMenuRef !== undefined &&
                    nextTabMenuRef.current !== null &&
                    (nextTabMenu.onClick !== undefined || onChange !== undefined)
                ) {
                    nextTabMenuRef.current.focus();
                    // Не совем удачное решение, переключать табики через onClick,
                    // тут могут возникнуть проблемы с event, т.к. в этом кейсе это KeyboardEvent.
                    if (nextTabMenu.onClick !== undefined) {
                        nextTabMenu.onClick(event as any);
                    }
                    if (onChange !== undefined) {
                        onChange(nextTabMenuIndex.toString());
                    }
                }
            }
        },
        [activeTab, tabsRefs, tabs, onChange],
    );

    return (
        <ul
            {...props}
            aria-orientation={orientation}
            className={cnTabsMenu(null, [className])}
            ref={innerRef}
            role="tablist"
        >
            {tabs.map(({ id, disabled, ...tabProps }, index) => (
                <TabsMenuTab
                    {...tabProps}
                    innerRef={tabsRefs[index]}
                    first={index === 0}
                    disabled={disabled}
                    onClick={disabled || (!onChange && !tabProps.onClick) ? undefined : (event) => {
                        if (tabProps.onClick !== undefined) {
                            tabProps.onClick(event);
                        }
                        if (onChange !== undefined) {
                            onChange(id as string);
                        }
                    }}
                    active={id === activeTab}
                    key={id}
                    onKeyDown={onKeyDown}
                />
            ))}
            {addonAfter}
        </ul>
    );
};

TabsMenu.displayName = cnTabsMenu();
