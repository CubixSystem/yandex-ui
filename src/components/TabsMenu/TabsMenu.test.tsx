import React, { createRef } from 'react';
import { mount, render } from 'enzyme';

import { delay, serverEnvironmentSetup } from '../../internal/utils';
import { jestEnvironmentSetup, getRoot } from '../../internal/utils/jestEnvironmentSetup';

import { TabsMenu } from './TabsMenu';
import { Keys } from '../../lib/keyboard';
import { ITabsMenuTabProps } from './Tab/TabsMenu-Tab';

const tab1: ITabsMenuTabProps = { id: 'tab1', onClick: jest.fn(), content: 'Поиск' };
const tab2: ITabsMenuTabProps = { id: 'tab2', onClick: jest.fn(), content: <div>Картинки</div> };
const tab3: ITabsMenuTabProps = { id: 'tab3', onClick: jest.fn(), disabled: true, content: 'Видео' };
const tab4: ITabsMenuTabProps = { id: 'tab4', onClick: jest.fn(), content: 'Карты' };

const tabs: ITabsMenuTabProps[] = [tab1, tab2, tab3, tab4];

describe('TabsMenu', () => {
    jestEnvironmentSetup();

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<TabsMenu activeTab={tab1.id} tabs={tabs} className="mix" />)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<TabsMenu activeTab={tab1.id} tabs={tabs} className="mix" />)).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            expect(TabsMenu.displayName).toBe('TabsMenu');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLUListElement>();
            mount(<TabsMenu activeTab={tab1.id} tabs={tabs} innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('активный таб должен иметь модификатор active', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(0)
                    .hasClass('TabsMenu-Tab_active'),
            ).toBe(true);
        });

        test('первый таб должен иметь модификатор first', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(0)
                    .hasClass('TabsMenu-Tab_first'),
            ).toBe(true);
        });

        test('disabled таб должен иметь модификатор disabled', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(2)
                    .hasClass('TabsMenu-Tab_disabled'),
            ).toBe(true);
        });

        test('должен менять активный таб при изменении свойства activeTab', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            wrapper.setProps({ activeTab: tab2.id });
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(1)
                    .hasClass('TabsMenu-Tab_active'),
            ).toBe(true);
        });

        test('должен вызывать onClick, если обработчик задан и таб не disabled', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            wrapper
                .find('.TabsMenu-Tab')
                .at(1)
                .simulate('click');
            expect(tab2.onClick).toBeCalledTimes(1);
        });

        test('должен вызывать onClick с правильными аргументами', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            wrapper
                .find('.TabsMenu-Tab')
                .at(3)
                .simulate('click');
            const onClickMock = tab4.onClick as jest.Mock;
            expect(onClickMock.mock.calls[0][0].target.innerHTML).toEqual(tab4.content);
        });

        test('не должен вызывать onClick, если tab disabled', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />);
            wrapper
                .find('.TabsMenu-Tab')
                .at(1)
                .simulate('click');
            expect(tab3.onClick).toBeCalledTimes(0);
        });

        test('должен вызывать onClick и выставлять фокус на следующий доступный пункт меню через клавиатурное управление', () => {
            const wrapper = mount(<TabsMenu activeTab={tab1.id} tabs={tabs} />, { attachTo: getRoot() });

            wrapper
                .find('.TabsMenu-Tab')
                .at(0)
                .simulate('keydown', { keyCode: Keys.RIGHT });
            expect(tab2.onClick).toBeCalledTimes(1);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(1)
                    .getDOMNode(),
            ).toBe(document.activeElement);

            wrapper.setProps({ activeTab: tab2.id });

            wrapper
                .find('.TabsMenu-Tab')
                .at(1)
                .simulate('keydown', { keyCode: Keys.DOWN });
            expect(tab4.onClick).toBeCalledTimes(1);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(3)
                    .getDOMNode(),
            ).toBe(document.activeElement);

            wrapper.setProps({ activeTab: tab4.id });

            wrapper
                .find('.TabsMenu-Tab')
                .at(3)
                .simulate('keydown', { keyCode: Keys.LEFT });
            expect(tab2.onClick).toBeCalledTimes(2);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(1)
                    .getDOMNode(),
            ).toBe(document.activeElement);

            wrapper.setProps({ activeTab: tab2.id });

            wrapper
                .find('.TabsMenu-Tab')
                .at(1)
                .simulate('keydown', { keyCode: Keys.UP });
            expect(tab1.onClick).toBeCalledTimes(1);
            expect(
                wrapper
                    .find('.TabsMenu-Tab')
                    .at(0)
                    .getDOMNode(),
            ).toBe(document.activeElement);
        });
    });
});
