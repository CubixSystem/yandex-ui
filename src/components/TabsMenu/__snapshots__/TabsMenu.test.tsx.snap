// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsMenu client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<ul
  aria-orientation="vertical"
  class="TabsMenu mix"
  role="tablist"
>
  <li
    aria-selected="true"
    class="TabsMenu-Tab TabsMenu-Tab_active TabsMenu-Tab_first"
    role="tab"
    tabindex="0"
  >
    Поиск
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab"
    role="tab"
    tabindex="-1"
  >
    <div>
      Картинки
    </div>
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab TabsMenu-Tab_disabled"
    role="tab"
  >
    Видео
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab"
    role="tab"
    tabindex="-1"
  >
    Карты
  </li>
</ul>
`;

exports[`TabsMenu server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<ul
  aria-orientation="vertical"
  class="TabsMenu mix"
  role="tablist"
>
  <li
    aria-selected="true"
    class="TabsMenu-Tab TabsMenu-Tab_active TabsMenu-Tab_first"
    role="tab"
    tabindex="0"
  >
    Поиск
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab"
    role="tab"
    tabindex="-1"
  >
    <div>
      Картинки
    </div>
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab TabsMenu-Tab_disabled"
    role="tab"
  >
    Видео
  </li>
  <li
    aria-selected="false"
    class="TabsMenu-Tab"
    role="tab"
    tabindex="-1"
  >
    Карты
  </li>
</ul>
`;
